src/box/lua/config/applier/roles.lua

```diff
@@ -0,0 +1,62 @@
+local log = require('internal.config.utils.log')
+
+local last_loaded = {}
+
+local function stop_roles(roles)
+    for role_name, role in pairs(last_loaded) do
+        if roles == nil or not roles[role_name] then
+            log.verbose('roles.post_apply: stop role ' .. role_name)
+            role.stop()
+        end
+    end
+end
+
+local function apply(config)
+    local configdata = config._configdata
+    local role_names = configdata:get('roles', {use_default = true})
+    if role_names == nil or next(role_names) == nil then
+        stop_roles()
+        return
+    end
+
+    -- Remove duplicates.
+    local roles = {}
+    for _, role_name in pairs(role_names) do
+        roles[role_name] = true
+    end
+
+    -- Run roles.
+    local roles_cfg = configdata:get('roles_cfg', {use_default = true}) or {}
+    local loaded = {}
+    for role_name in pairs(roles) do
+        local role = last_loaded[role_name]
+        if not role then
+            role = require(role_name)
+            local methods = {'init', 'validate_config', 'apply_config', 'stop'}
+            for _, method_name in pairs(methods) do
+                if type(role[method_name]) ~= 'function' then
+                    local err = 'Role %s does not contain method %s'
```

Comment:
Nit: a method assumes that it receives its own object as a first argument. I would call them functions.

Reply:
Fixed.